pkg/buildpack/managed_collection.go

```diff
@@ -1,60 +1,144 @@
 package buildpack
 
-type ManagedCollection struct {
+// ManagedCollection defines the required behavior to deal with build modules when adding then to an OCI image.
+type ManagedCollection interface {
+	// AllModules returns all build modules handle by the manager
+	AllModules() []BuildModule
+
+	// ExplodedModules returns all build modules that will be added to the output artifact as a single layer
+	// containing a single module.
+	ExplodedModules() []BuildModule
+
+	// AddModules determines whether the explodedModules must be added as flattened or not.
+	AddModules(main BuildModule, deps ...BuildModule)
+
+	// FlattenedModules returns all build modules that will be added to the output artifact as a single layer
+	// containing multiple modules.
+	FlattenedModules() [][]BuildModule
+
+	// ShouldFlatten returns true if the given module should be flattened.
+	ShouldFlatten(module BuildModule) bool
+}
+
+type managedCollection struct {
 	explodedModules  []BuildModule
-	flattenedModules []BuildModule
-	flatten          bool
+	flattenedModules [][]BuildModule
 }
 
-func NewModuleManager(flatten bool) *ManagedCollection {
-	return &ManagedCollection{
-		flatten:          flatten,
-		explodedModules:  []BuildModule{},
-		flattenedModules: []BuildModule{},
-	}
+func (f *managedCollection) ExplodedModules() []BuildModule {
+	return f.explodedModules
+}
+
+func (f *managedCollection) FlattenedModules() [][]BuildModule {
+	return f.flattenedModules
 }
 
-// AllModules returns all explodedModules handle by the manager
-func (f *ManagedCollection) AllModules() []BuildModule {
+func (f *managedCollection) AllModules() []BuildModule {
 	all := f.explodedModules
-	all = append(all, f.flattenedModules...)
+	for _, modules := range f.flattenedModules {
+		all = append(all, modules...)
+	}
 	return all
 }
 
-// ExplodedModules returns all modules that will be added to the output artifact as a single layer containing a single module.
-func (f *ManagedCollection) ExplodedModules() []BuildModule {
-	return f.explodedModules
+func (f *managedCollection) ShouldFlatten(module BuildModule) bool {
+	for _, modules := range f.flattenedModules {
+		for _, v := range modules {
+			if v == module {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// managedCollectionV1 can be used to flatten all the flattenModuleInfos or none of them.
+type managedCollectionV1 struct {
+	managedCollection
+	flattenAll bool
 }
 
-// FlattenedModules returns all modules that will be added to the output artifact as a single layer containing multiple modules.
-func (f *ManagedCollection) FlattenedModules() [][]BuildModule {
-	if f.flatten {
-		modules := [][]BuildModule{}
-		modules = append(modules, f.flattenedModules)
-		return modules
+func NewManagedCollectionV1(flattenAll bool) ManagedCollection {
```

Comment:
Could we add a docs comment here? In particular, it could be helpful to know that `NewManagedCollectionV1` is for buildpack (package) builders, and `NewManagedCollectionV2` is for builder builders.